File: Java/src/ArmorBuilder.java
package net.glitchtechs.starapi;

import java.util.ArrayList;
import java.util.List;
import org.bukkit.Bukkit;
import org.bukkit.Color;
import org.bukkit.Material;
import org.bukkit.attribute.Attribute;
import org.bukkit.attribute.AttributeModifier;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.inventory.ItemFlag;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;
import org.bukkit.inventory.meta.LeatherArmorMeta;

public class ArmorBuilder {
    private ItemStack armorPiece;

    public ArmorBuilder(Material armorType) {
        this.armorPiece = new ItemStack(armorType);
    }

    public ArmorBuilder displayName(String displayName) {
        this.setItemDisplayName(displayName);
        return this;
    }

    public ArmorBuilder lore(String... lore) {
        this.setItemLore(lore);
        return this;
    }

    public ArmorBuilder amount(int amount) {
        this.setItemAmount(amount);
        return this;
    }

    public ArmorBuilder enchant(Enchantment enchantment, int level) {
        this.enchantItem(enchantment, level);
        return this;
    }

    public ArmorBuilder attribute(Attribute attribute, AttributeModifier modifier) {
        this.setItemAttribute(attribute, modifier);
        return this;
    }

    public ArmorBuilder flags(ItemFlag... flags) {
        this.setItemFlags(flags);
        return this;
    }

    public ArmorBuilder leatherColor(Color color) {
        if (this.armorPiece.getType() == Material.LEATHER_HELMET || this.armorPiece.getType() == Material.LEATHER_CHESTPLATE || this.armorPiece.getType() == Material.LEATHER_LEGGINGS || this.armorPiece.getType() == Material.LEATHER_BOOTS) {
            LeatherArmorMeta leatherMeta = (LeatherArmorMeta)this.getItemMeta();
            leatherMeta.setColor(color);
            this.armorPiece.setItemMeta(leatherMeta);
        }

        return this;
    }

    public ItemStack build() {
        return this.armorPiece;
    }

    private void setItemDisplayName(String displayName) {
        ItemMeta itemMeta = this.getItemMeta();
        itemMeta.setDisplayName(displayName);
        this.armorPiece.setItemMeta(itemMeta);
    }

    private void setItemLore(String... lore) {
        ItemMeta itemMeta = this.getItemMeta();
        List<String> loreList = new ArrayList();
        String[] var4 = lore;
        int var5 = lore.length;

        for(int var6 = 0; var6 < var5; ++var6) {
            String line = var4[var6];
            loreList.add(line);
        }

        itemMeta.setLore(loreList);
        this.armorPiece.setItemMeta(itemMeta);
    }

    private void setItemAmount(int amount) {
        this.armorPiece.setAmount(amount);
    }

    private void enchantItem(Enchantment enchantment, int level) {
        ItemMeta itemMeta = this.getItemMeta();
        itemMeta.addEnchant(enchantment, level, true);
        this.armorPiece.setItemMeta(itemMeta);
    }

    private void setItemAttribute(Attribute attribute, AttributeModifier modifier) {
        ItemMeta itemMeta = this.getItemMeta();
        itemMeta.addAttributeModifier(attribute, modifier);
        this.armorPiece.setItemMeta(itemMeta);
    }

    private void setItemFlags(ItemFlag... flags) {
        ItemMeta itemMeta = this.getItemMeta();
        itemMeta.addItemFlags(flags);
        this.armorPiece.setItemMeta(itemMeta);
    }

    private ItemMeta getItemMeta() {
        ItemMeta itemMeta = this.armorPiece.getItemMeta();
        if (itemMeta == null) {
            itemMeta = Bukkit.getItemFactory().getItemMeta(this.armorPiece.getType());
            this.armorPiece.setItemMeta(itemMeta);
        }

        return itemMeta;
    }
}

File: Kotlin/src/ArmorBuilder.kt
package net.glitchtechs.starapi

import org.bukkit.Bukkit
import org.bukkit.Color
import org.bukkit.Material
import org.bukkit.attribute.Attribute
import org.bukkit.attribute.AttributeModifier
import org.bukkit.enchantments.Enchantment
import org.bukkit.inventory.ItemFlag
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.ItemMeta
import org.bukkit.inventory.meta.LeatherArmorMeta


class ArmorBuilder(armorType: Material?) {
    private val armorPiece = ItemStack(armorType!!)

    fun displayName(displayName: String): ArmorBuilder {
        this.setItemDisplayName(displayName)
        return this
    }

    fun lore(vararg lore: String?): ArmorBuilder {
        this.setItemLore(*lore as Array<String>)
        return this
    }

    fun amount(amount: Int): ArmorBuilder {
        this.setItemAmount(amount)
        return this
    }

    fun enchant(enchantment: Enchantment, level: Int): ArmorBuilder {
        this.enchantItem(enchantment, level)
        return this
    }

    fun attribute(attribute: Attribute, modifier: AttributeModifier): ArmorBuilder {
        this.setItemAttribute(attribute, modifier)
        return this
    }

    fun flags(vararg flags: ItemFlag?): ArmorBuilder {
        this.setItemFlags(*flags as Array<ItemFlag>)
        return this
    }

    fun leatherColor(color: Color?): ArmorBuilder {
        if (armorPiece.type == Material.LEATHER_HELMET || (armorPiece.type == Material.LEATHER_CHESTPLATE) || (armorPiece.type == Material.LEATHER_LEGGINGS) || (armorPiece.type == Material.LEATHER_BOOTS)) {
            val leatherMeta = itemMeta as LeatherArmorMeta?
            leatherMeta!!.setColor(color)
            armorPiece.setItemMeta(leatherMeta)
        }

        return this
    }

    fun build(): ItemStack {
        return this.armorPiece
    }

    private fun setItemDisplayName(displayName: String) {
        val itemMeta = this.itemMeta
        itemMeta!!.setDisplayName(displayName)
        armorPiece.setItemMeta(itemMeta)
    }

    private fun setItemLore(vararg lore: String) {
        val itemMeta = this.itemMeta ?: return
        val loreList: MutableList<String> = ArrayList()

        for (line in lore) {
            loreList.add(line)
        }

        itemMeta.lore = loreList
        armorPiece.itemMeta = itemMeta
    }


    private fun setItemAmount(amount: Int) {
        armorPiece.amount = amount
    }

    private fun enchantItem(enchantment: Enchantment, level: Int) {
        val itemMeta = this.itemMeta
        itemMeta!!.addEnchant(enchantment, level, true)
        armorPiece.setItemMeta(itemMeta)
    }

    private fun setItemAttribute(attribute: Attribute, modifier: AttributeModifier) {
        val itemMeta = this.itemMeta
        itemMeta!!.addAttributeModifier(attribute, modifier)
        armorPiece.setItemMeta(itemMeta)
    }

    private fun setItemFlags(vararg flags: ItemFlag) {
        val itemMeta = this.itemMeta
        itemMeta!!.addItemFlags(*flags)
        armorPiece.setItemMeta(itemMeta)
    }

    private val itemMeta: ItemMeta?
        get() {
            var itemMeta = armorPiece.itemMeta
            if (itemMeta == null) {
                itemMeta = Bukkit.getItemFactory().getItemMeta(armorPiece.type)
                armorPiece.setItemMeta(itemMeta)
            }

            return itemMeta
        }
}